// models/models.js
'use strict';
const qtoolsGen = require('qtools');
const qtools = new qtoolsGen(module);
const async = require('async');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {

	qtools.validateProperties({
		subject: args || {},
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'config',
				optional: false
			},
			{
				name: 'apiManager',
				optional: false
			},
			{
				name: 'router',
				optional: false
			},
			{
				name: 'permissionMaster',
				optional: false
			},
			{
				name: 'initCallback',
				optional: false
			}
		]
	});

	//LOCAL VARIABLES ====================================
	
	const mongoose = require('mongoose');
	
	mongoose.connect(this.config.database.connectionString).connection;

	let workerList = {};

	//LOCAL FUNCTIONS ====================================

	const startSystem = () => {

		const startList = [];


		const users = require('users');
		startList.push((done) => {
			const workerName = 'users'
			new users({
				config: this.config,
				apiManager: this.apiManager.init(workerName),
				router: this.router,
				permissionMaster: this.permissionMaster,
				mongoose:mongoose,
				initCallback: function() {
					workerList[workerName] = this; done();
				}
			});
		});


		const bookmarkGrids = require('bookmarks');
		startList.push((done) => {
			const workerName = 'bookmarks'
			new bookmarkGrids({
				config: this.config,
				apiManager: this.apiManager.init(workerName),
				router: this.router,
				permissionMaster: this.permissionMaster,
				mongoose:mongoose,
				initCallback: function() {
					workerList[workerName] = this; done();
				}
			});
		});

		async.series(startList, () => {
			this.initCallback && this.initCallback();
		});
	};

	//METHODS AND PROPERTIES ====================================
	if (typeof (workerList) != 'object') {
		this.shutdown = (message, callback) => {
			console.log(`\nshutting down ${__dirname}`);
			callback('', message);
		}
	} else {
		const buildShutdownList = (message) => {
			const shutdownList = [];
			for (var i in workerList) {
				var worker = workerList[i];
				shutdownList.push(
					((i) => {
						return (done) => {
							workerList[i].shutdown(message, done)
						}
					})(i)
				);
			}
			return shutdownList;
		};

		const cleanup = () => {
			let nameString = '';
			for (var i in workerList) {
				workerList[i] = null;
				nameString += `${i}, `;
			}
			qtools.message(`[${nameString.replace(/, $/, '')}] were flushed at ${Date.now()}`);
			workerList = {};
		}

		this.shutdown = (message, callback) => {
			async.parallel(buildShutdownList(message), () => {
				cleanup();
				callback('', message);
			});
		}
	}

	//API ENDPOINTS ====================================

	let route;
	let method;
	
	if (qtools.getSurePath(this, 'config.system.environment')=='development'){
		route = new RegExp('/'+require('path').parse(module.id).name+'$');
		method='get';
		this.permissionMaster.addRoute(method, route, 'all');
		this.router[method](route, (req, res, next)=>{
			res.send(`webInit() says, ${this.config.webInit.name} is up and running at ${req.path}`);
		})
	}
	
	//INITIALIZATION ====================================

	startSystem();

	return this;
};

//END OF moduleFunction() ============================================================

module.exports = moduleFunction;
//module.exports = new moduleFunction();
